File: voice_typer/client/src/main/single_instance.ts
/**
 * Single-instance gate + stale-lock recovery.
 *
 * Extracted from `index.ts` (REF-2). Owns:
 *   - `computeConfigDir()` — shared config-directory resolver (mirrors
 *     the Python backend's `_config_dir()` in `voice_typer/server/config.py`).
 *   - `electronPidFile()` / `writeElectronPidFile()` / `clearElectronPidFile()`
 *     / `readStaleElectronPid()` — PID-file management for stale-lock
 *     detection.
 *   - `acquireSingleInstanceLock()` — top-level gate that must run
 *     before `app.whenReady()`. On a duplicate launch it calls
 *     `app.exit(0)`; on the primary instance it writes the PID file and
 *     registers `app.on("second-instance", …)` to show + focus the
 *     dashboard window.
 */
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { app } from "electron";
import { log } from "./logging";
import { showMainWindow } from "./windows";

/**
 * Compute the config dir the same way `app.whenReady()` does so the
 * `electron.pid` file lives alongside the `backend.pid` file written by
 * Python.
 *
 * Mirrors `_config_dir()` in `voice_typer/server/config.py`:
 *   1. `VOICE_TYPER_CONFIG_DIR` env var (if set)
 *   2. Legacy `~/.voice-typer` (if it exists — migration path)
 *   3. Platform-appropriate path (`%APPDATA%/voice-typer` on Windows,
 *      `~/Library/Application Support/voice-typer` on macOS,
 *      `$XDG_DATA_HOME/voice-typer` on Linux)
 */
export function computeConfigDir(): string {
	const envOverride = process.env.VOICE_TYPER_CONFIG_DIR;
	if (envOverride) return envOverride;
	const legacy = path.join(os.homedir(), ".voice-typer");
	try {
		if (fs.existsSync(legacy)) return legacy;
	} catch (e) {
		// ignore — fs.existsSync may throw on permission denied or
		// if the homedir is unreachable. Fall through to the
		// platform-appropriate default path below.
		log.warn("[single_instance] legacy config dir probe failed:", e);
	}
	if (process.platform === "win32") {
		return path.join(process.env.APPDATA || os.homedir(), "voice-typer");
	}
	if (process.platform === "darwin") {
		return path.join(
			os.homedir(),
			"Library",
			"Application Support",
			"voice-typer",
		);
	}
	return path.join(
		process.env.XDG_DATA_HOME || path.join(os.homedir(), ".local", "share"),
		"voice-typer",
	);
}

export function electronPidFile(): string {
	return path.join(computeConfigDir(), "electron.pid");
}

export function writeElectronPidFile(): void {
	try {
		const dir = path.dirname(electronPidFile());
		fs.mkdirSync(dir, { recursive: true });
		fs.writeFileSync(electronPidFile(), `${process.pid}\n`, {
			encoding: "utf-8",
			flag: "w",
			mode: 0o600,
		});
	} catch (e) {
		// best-effort: PID file is only used for stale-lock detection
		// on the next launch; a missing file just means a future launch
		// can't auto-recover from a hard crash (it will fall back to
		// exiting as a duplicate instance).
		log.warn("[single_instance] writeElectronPidFile failed:", e);
	}
}

export function clearElectronPidFile(): void {
	try {
		if (fs.existsSync(electronPidFile())) fs.unlinkSync(electronPidFile());
	} catch (e) {
		// best-effort: file may already be gone (race with another
		// process) or the FS may be read-only. Non-fatal — a leftover
		// PID file just means the next launch does stale-lock detection.
		log.warn("[single_instance] clearElectronPidFile failed:", e);
	}
}

/**
 * Verify the process at ``pid`` is plausibly the Voice Typer Electron
 * process (not an unrelated process that happened to reuse the PID
 * after a crash).
 *
 * On Linux, reads ``/proc/<pid>/cmdline`` and checks for an Electron /
 * Voice Typer marker (``electron`` or ``voice-typer``). On macOS, runs
 * ``ps -p <pid> -o comm=`` and applies the same check. On Windows,
 * runs ``wmic process where processid=<pid> get commandline`` and
 * applies the same check.
 *
 * Returns ``true`` if the PID exists AND the process command line
 * matches Voice Typer; ``false`` otherwise. Any error (file missing,
 * spawn failure, permission denied) returns ``false`` so the caller
 * falls back to the conservative "PID is alive" path (treats the PID
 * as still-held by Voice Typer — preventing accidental lockout of an
 * unrelated process).
 */
export function isPidVoiceTyper(pid: number): boolean {
	try {
		let cmdline = "";
		if (process.platform === "linux") {
			cmdline = fs
				.readFileSync(`/proc/${pid}/cmdline`, "utf-8")
				.replace(/\0/g, " ");
		} else if (process.platform === "darwin") {
			const { execSync } =
				require("node:child_process") as typeof import("node:child_process");
			cmdline = execSync(`ps -p ${pid} -o comm=`, { encoding: "utf-8" });
		} else if (process.platform === "win32") {
			const { execSync } =
				require("node:child_process") as typeof import("node:child_process");
			cmdline = execSync(
				`wmic process where processid=${pid} get commandline /value`,
				{ encoding: "utf-8" },
			);
		} else {
			return true;
		}
		const lower = cmdline.toLowerCase();
		return (
			lower.includes("electron") ||
			lower.includes("voice-typer") ||
			lower.includes("voice_typer")
		);
	} catch (e) {
		// Any read/spawn failure (file missing, permission denied,
		// ps/wmic not on PATH) returns false so the caller falls back
		// to the conservative "PID is alive" path. Previously this was
		// a silent catch — a flaky /proc mount or a stripped-down
		// Windows image that lacked wmic would silently mis-classify
		// every stale-PID check. Log the failure so operators can
		// diagnose the root cause from the runtime log.
		log.warn("[single_instance] isPidVoiceTyper failed:", e);
		return false;
	}
}

/**
 * Read the PID file and return the PID if the process it points to is DEAD
 * (stale lock).  Returns null if the file doesn't exist, is unreadable, or
 * the PID is still alive.
 *
 * : a stale-PID check via ``process.kill(pid, 0)`` only verifies
 * the PID exists — not that it's Voice Typer. PID reuse by an unrelated
 * process would cause a lockout until that process exits. We now also
 * verify the process command line via :func:`isPidVoiceTyper`.
 */
export function readStaleElectronPid(): number | null {
	try {
		if (!fs.existsSync(electronPidFile())) return null;
		const content = fs.readFileSync(electronPidFile(), "utf-8").trim();
		if (!content) return null;
		const pid = parseInt(content, 10);
		if (!Number.isFinite(pid) || pid <= 0) return null;
		try {
			process.kill(pid, 0);
			//verify it's actually Voice Typer before
			// treating the lock as held. If the PID was reused by
			// an unrelated process, treat the lock as stale so we
			// don't lock out the unrelated process.
			if (!isPidVoiceTyper(pid)) {
				log.warn(
					`[single_instance] PID ${pid} is alive but is not Voice Typer ` +
						"(PID reuse) — treating lock as stale",
				);
				return pid;
			}
			return null; // still alive AND is Voice Typer
		} catch (e) {
			// process.kill(pid, 0) threw — ESRCH (PID is gone) is
			// the EXPECTED signal that the previous Voice Typer
			// instance crashed hard. Other errors (EPERM on a PID
			// owned by another user, EINVAL on a malformed pid)
			// are rarer and worth surfacing. Log the signal so the
			// stale-lock recovery path is diagnosable instead of
			// opaque (the only observable was the return value).
			log.warn(
				`[single_instance] readStaleElectronPid process.kill(${pid}, 0) threw — treating as stale:`,
				e,
			);
			return pid; // stale — process is gone
		}
	} catch (e) {
		// Outer fallback: fs.existsSync / fs.readFileSync on the PID
		// file itself failed (permission denied, FS read-only, file
		// vanished mid-read). Returns null so the caller treats the
		// lock as held (conservative — don't accidentally double-launch).
		// Previously this was a silent catch — a corrupt PID file or
		// a transient FS error was invisible. Log so operators can
		// see when the gate is falling back to its safe default.
		log.warn("[single_instance] readStaleElectronPid PID-file read failed:", e);
		return null;
	}
}

/**
 * Acquire the single-instance lock with at most one stale-lock retry.
 *
 * MUST run before `app.whenReady()` — the lock is checked at process start.
 *
 * Behaviour:
 *   - If the first `requestSingleInstanceLock()` fails AND the existing
 *     PID is stale (previous instance crashed hard), release the stale
 *     lock and retry once.
 *   - If still no lock OR `VT_FOCUS_ONLY=1`, call `app.exit(0)` to
 *     terminate immediately (the primary instance has already received
 *     or is about to receive the `second-instance` event and will show
 *     itself).
 *   - Otherwise (we got the lock), write our PID file and register
 *     `app.on("second-instance", …)` to show + focus the dashboard.
 *
 * `VT_FOCUS_ONLY` is set by `autostart_launcher._focus_running_app()`
 * as a defensive guard: if this env var is set, this process is a
 * lightweight duplicate that must exit without doing ANY heavy init.
 */
export function acquireSingleInstanceLock(): void {
	let gotTheLock = app.requestSingleInstanceLock();
	if (!gotTheLock) {
		const stalePid = readStaleElectronPid();
		if (stalePid !== null) {
			log.warn(
				`[STARTUP] single-instance lock held by dead PID ${stalePid} — ` +
					"clearing stale PID file and retrying",
			);
			clearElectronPidFile();
			//``app.releaseSingleInstanceLock()`` releases the lock
			// held by THIS process. We never held the lock (the first
			// ``requestSingleInstanceLock()`` returned false), so this call
			// is a no-op on most platforms. On Linux, Electron's single-
			// instance lock is a ``SingletonLock`` symlink in the userData
			// directory; ``releaseSingleInstanceLock`` deletes it. If the
			// dead process's lock file is still on disk, we manually
			// delete it as a defense-in-depth measure so the retry has a
			// clean slate. On Windows, the named mutex is auto-released
			// when the owning process exits (no manual cleanup needed).
			// On macOS, the file lock is released by the OS on process
			// exit (no manual cleanup needed).
			if (process.platform === "linux") {
				try {
					const userDataPath = app.getPath("userData");
					const singletonLock = path.join(userDataPath, "SingletonLock");
					if (fs.existsSync(singletonLock)) {
						fs.unlinkSync(singletonLock);
						log.warn(
							`[single_instance] deleted stale Linux SingletonLock symlink at ${singletonLock}`,
						);
					}
				} catch (e) {
					/* best-effort — the retry below will fail if the lock is still held */
					log.warn("[single_instance] Linux SingletonLock cleanup failed:", e);
				}
			}
			try {
				app.releaseSingleInstanceLock();
			} catch (e) {
				/* ignore — Electron may reject if we never held the lock */
				log.warn("[single_instance] releaseSingleInstanceLock failed:", e);
			}
			gotTheLock = app.requestSingleInstanceLock();
			//log the result of the retry so operators can see
			// whether the stale-lock recovery succeeded. Previously the
			// result was silently assigned to ``gotTheLock`` and the only
			// observable signal was whether the process exited (duplicate)
			// or continued (primary). A retry that failed but didn't exit
			// (e.g. because VT_FOCUS_ONLY was not set) would silently run
			// as a "ghost" instance with no lock.
			log.warn(
				`[single_instance] retry requestSingleInstanceLock() returned ${gotTheLock}`,
			);
		}
	}
	if (!gotTheLock || process.env.VT_FOCUS_ONLY === "1") {
		// We are the duplicate (or a focus-only probe).  The first instance has
		// already received (or is about to receive) the "second-instance" event
		// and will show itself.
		//
		// Use app.exit(0) instead of app.quit() to guarantee immediate
		// termination — app.quit() allows the event loop to drain (which can
		// fire whenReady and start Python before the process exits), while
		// app.exit(0) terminates without waiting.
		app.exit(0);
	} else {
		// P1-1.4: we got the lock — write our PID so a future launch can
		// detect if we've crashed hard and release the stale lock.
		writeElectronPidFile();
		app.on("second-instance", () => {
			// Another launch attempt happened.  Show + focus the dashboard so it
			// feels like the app "opened."  Create it lazily if autostart started
			// us hidden and the user never opened it yet.
			showMainWindow();
		});
	}
}

File: voice_typer/worker/_single_instance.py
"""Single-instance lock for the worker process (master plan §7.2).

This module is an intentional extraction from ``voice_typer/worker/__main__.py``
per E3 (no spaghetti entry files). It mirrors
:mod:`voice_typer.server.single_instance`'s shape: a lock file in the
canonical app config dir, ``O_CREAT | O_EXCL | O_CLOEXEC`` + ``flock``
on POSIX, best-effort existence check on Windows, and stale-PID
recovery when the holder process is dead.

The worker single-instance lock file name (``worker.lock``) is distinct
from the slim-core sidecar's ``backend.lock`` so the two processes can
run side-by-side (master plan §7.1 "1-host ↔ 2-processes pattern").
"""

from __future__ import annotations

import contextlib
import logging
import os
from pathlib import Path

log = logging.getLogger("voice_typer.worker")

# Worker single-instance lock file name. Distinct from the slim-core
# sidecar's ``backend.lock`` so the two processes can run side-by-side
# (master plan §7.1 "1-host ↔ 2-processes pattern"). Lives in the
# canonical app config dir so it's resolved per-platform (Windows:
# ``%APPDATA%/voice-typer``, macOS: ``~/Library/Application
# Support/voice-typer``, Linux: ``$XDG_DATA_HOME/voice-typer`` —
# resolved via :func:`voice_typer.server.config._config_dir`).
_WORKER_LOCK_NAME = "worker.lock"


def _worker_lock_path() -> Path:
    """Resolve the worker single-instance lock file path.

    Uses :func:`voice_typer.server.config._config_dir` (the canonical
    per-platform app data dir) so the lock file lives under the same
    O3 ``run/`` subdir as the slim-core sidecar's ``backend.lock`` —
    same dir, different file name, so the two processes do not contend
    on the same lock.
    """
    from voice_typer.server._paths import RUN_SUBDIR
    from voice_typer.server.config import _config_dir

    return _config_dir() / RUN_SUBDIR / _WORKER_LOCK_NAME


class _WorkerSingleInstanceHandle:
    """POSIX single-instance lock handle for the worker.

    Mirrors :class:`voice_typer.server.single_instance._PosixSingleInstanceHandle`:
    the fd is held for the process lifetime, ``release()`` closes it and
    unlinks the lockfile (idempotent, best-effort).

    On Windows the lock is best-effort (no named mutex is used here —
    the worker is always spawned by the Tauri host, which already
    enforces single-instance via ``tauri-plugin-single-instance``; the
    Python-side lock is defense-in-depth for dev-runs from a terminal).
    """

    __slots__ = ("_fd", "_path", "_released")

    def __init__(self, fd: int, path: Path) -> None:
        self._fd = fd
        self._path = path
        self._released = False

    def release(self) -> None:
        """Close the lockfile fd (POSIX) / unlink the lockfile (best-effort).

        Idempotent: subsequent calls are no-ops. Safe to call after the
        underlying fd has already been closed by other means (errors
        from ``os.close`` / ``os.unlink`` are suppressed at DEBUG
        level).
        """
        if self._released:
            return
        self._released = True
        if self._fd >= 0:
            with contextlib.suppress(OSError):
                os.close(self._fd)
        with contextlib.suppress(OSError):
            self._path.unlink(missing_ok=True)


def _ensure_worker_single_instance() -> _WorkerSingleInstanceHandle | None:
    """Acquire the worker single-instance lock.

    Returns a handle whose ``release()`` method releases the lock (call
    on shutdown). On duplicate launch (POSIX: ``O_EXCL`` fails; Windows:
    existence check), logs at WARNING and returns ``None`` — the caller
    decides whether to exit.

    Stale-PID recovery (POSIX): if the lockfile exists but the PID
    inside is not alive, the lockfile is reclaimed (mirrors
    :func:`voice_typer.server.single_instance._ensure_single_instance_posix`'s
    stale-PID path).
    """
    lock_path = _worker_lock_path()
    try:
        lock_path.parent.mkdir(parents=True, exist_ok=True)
    except OSError:
        log.debug("[WORKER] could not create lockfile parent dir — single-instance is best-effort", exc_info=True)
        return None

    if os.name == "posix":
        import fcntl  # POSIX-only stdlib

        # O_CREAT | O_EXCL | O_CLOEXEC: primary mechanism. If the file
        # already exists, we fall through to the stale-PID recovery
        # path (mirrors ``_ensure_single_instance_posix``).
        try:
            fd = os.open(lock_path, os.O_CREAT | os.O_EXCL | os.O_WRONLY | os.O_CLOEXEC, 0o600)
            try:
                os.write(fd, f"{os.getpid()}\n".encode("ascii"))
            except OSError:
                log.debug("[WORKER] failed to write PID to lockfile — single-instance is best-effort", exc_info=True)
            # flock as defense-in-depth (mirrors single_instance.py).
            with contextlib.suppress(OSError):
                fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
            return _WorkerSingleInstanceHandle(fd, lock_path)
        except FileExistsError:
            # Stale-PID recovery: read the PID, check liveness.
            try:
                pid_str = lock_path.read_text(encoding="ascii").strip()
                pid = int(pid_str)
            except (OSError, ValueError):
                log.warning("[WORKER] worker.lock exists but is unreadable — refusing to start (duplicate instance?)")
                return None
            # Check liveness via os.kill(pid, 0). On POSIX this returns
            # None if the process is alive, raises ProcessLookupError if
            # it's dead.
            try:
                os.kill(pid, 0)
                log.warning("[WORKER] worker already running (pid=%d) — refusing to start", pid)
                return None
            except ProcessLookupError:
                # Stale lockfile — reclaim it.
                log.info("[WORKER] reclaiming stale worker.lock (pid=%d was dead)", pid)
                with contextlib.suppress(OSError):
                    lock_path.unlink(missing_ok=True)
                # Retry once.
                try:
                    fd = os.open(lock_path, os.O_CREAT | os.O_EXCL | os.O_WRONLY | os.O_CLOEXEC, 0o600)
                    with contextlib.suppress(OSError):
                        os.write(fd, f"{os.getpid()}\n".encode("ascii"))
                    with contextlib.suppress(OSError):
                        fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
                    return _WorkerSingleInstanceHandle(fd, lock_path)
                except OSError:
                    log.warning("[WORKER] could not reclaim worker.lock — refusing to start", exc_info=True)
                    return None
            except PermissionError:
                # PID is alive but owned by another user (rare).
                log.warning("[WORKER] worker.lock held by pid=%d (permission check) — refusing to start", pid)
                return None
    else:
        # Windows: best-effort existence check. The Tauri host's
        # ``tauri-plugin-single-instance`` is the authoritative gate;
        # this is defense-in-depth for dev-runs from a terminal.
        if lock_path.exists():
            try:
                pid_str = lock_path.read_text(encoding="ascii").strip()
                pid = int(pid_str)
                # On Windows there is no portable ``os.kill(pid, 0)`` —
                # we use the lockfile's existence as the signal. A stale
                # lockfile from a crashed worker is reclaimed below if
                # the PID's process tree is gone (checked via
                # ``os.kill``-equivalent on Windows; left as TODO since
                # the Tauri host owns authoritative single-instance).
                log.warning("[WORKER] worker.lock exists (pid=%d) — refusing to start", pid)
                return None
            except (OSError, ValueError):
                log.warning("[WORKER] worker.lock exists but is unreadable — refusing to start")
                return None
        try:
            lock_path.write_text(f"{os.getpid()}\n", encoding="ascii")
        except OSError:
            log.debug("[WORKER] could not write worker.lock — single-instance is best-effort", exc_info=True)
            return None
        # fd=-1 (no POSIX fd to close); release() will just unlink.
        return _WorkerSingleInstanceHandle(-1, lock_path)
